org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/BundleUtils.java

```diff
@@ -37,112 +37,141 @@
-		if (useReference) {
-			bundleLocation = REFERENCE_PREFIX + bundleLocation;
-		}
-		return bundleLocation;
-	}
-
-	private static BundleInfo getBundleInfo(String bundleLocation) throws IOException {
-		try (JarFile jarFile = new JarFile(bundleLocation)) {
-			Manifest manifest = jarFile.getManifest();
-			if (manifest != null) {
-				Attributes mainAttributes = manifest.getMainAttributes();
-				if (mainAttributes != null) {
-					String bundleVersion = mainAttributes.getValue(Constants.BUNDLE_VERSION);
-					if (StringUtils.isBlank(bundleVersion)) {
-						return null;
-					}
-					String symbolicName = mainAttributes.getValue(Constants.BUNDLE_SYMBOLICNAME);
-					if (StringUtils.isNotBlank(symbolicName) && symbolicName.indexOf(";") >= 0) {
-						symbolicName = symbolicName.substring(0, symbolicName.indexOf(";"));
-					}
-					return new BundleInfo(bundleVersion, symbolicName);
-				}
-			}
-		}
-		return null;
-	}
+  private static final String REFERENCE_PREFIX = "reference:";
+
+  private static class BundleInfo {
+
+    private String version;
+
+    private String symbolicName;
+
+    private BundleInfo(String bundleVersion, String symbolicName) {
+      this.version = bundleVersion;
+      this.symbolicName = symbolicName;
+    }
+
+    private String getVersion() {
+      return version;
+    }
+
+    private String getSymbolicName() {
+      return symbolicName;
+    }
+  }
+
+  /**
+   * Load a collection of bundle based on the provided file path locations.
+   *
+   * @param bundleLocations
+   *            The collection of the bundle file path location
+   * @throws CoreException
+   *             throw the <code>CoreException</code> if failed to load any
+   *             bundles.
+   */
+  public static void loadBundles(Collection<String> bundleLocations) throws CoreException {
+    if (bundleLocations == null || bundleLocations.isEmpty()) {
+      return;
+    }
+
+    BundleContext context = JavaLanguageServerPlugin.getBundleContext();
+    MultiStatus status = new MultiStatus(context.getBundle().getSymbolicName(), IStatus.OK, "Load bundle list", null);
+    Collection<Bundle> bundlesToStart = new ArrayList<>();
+    for (String bundleLocation : bundleLocations) {
+      try {
+        if (StringUtils.isEmpty(bundleLocation)) {
+          JavaLanguageServerPlugin.logError("Empty bundle location");
+          continue;
+        }
+
+        String location = getBundleLocation(bundleLocation, true);
+
+        BundleInfo bundleInfo = getBundleInfo(bundleLocation);
+        if (bundleInfo == null) {
+          status.add(new Status(IStatus.ERROR, context.getBundle().getSymbolicName(), "Failed to get bundleInfo for bundle from " + bundleLocation, null));
+          continue;
+        }
+        Bundle bundle = Platform.getBundle(bundleInfo.getSymbolicName());
+        if (bundle != null) {
+          if (bundle.getLocation().equals(location) && bundle.getVersion().equals(Version.parseVersion(bundleInfo.getVersion()))) {

```

Comment:
so  bundleInfo.getVersion() can be null at that point, which would make the Version constructor barf

Reply:
Not an issue since Version constructor is not used

Reply:
ok pretend I didn't say anything :-)